ft_otp/oath-toolkit-2.6.11/oathtool/oathtool.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.1.
.TH OATHTOOL "1" "January 2024" "OATH Toolkit 2.6.11" "User Commands"
.SH NAME
oathtool \- Open AuTHentication (OATH) one-time password tool
.SH SYNOPSIS
.B oathtool
[\fI\,OPTION\/\fR]... [\fI\,KEY \/\fR[\fI\,OTP\/\fR]]...
.SH DESCRIPTION
Generate and validate OATH one\-time passwords.  KEY and OTP is the string '\-'
to read from standard input, '@FILE' to read from indicated filename, or a hex
encoded value (not recommended on multi\-user systems).
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit
.TP
\fB\-\-hotp\fR
use event\-based HOTP mode  (default=on)
.TP
\fB\-\-totp\fR[=\fI\,MODE\/\fR]
use time\-variant TOTP mode (values "SHA1",
"SHA256", or "SHA512")  (default=`SHA1')
.TP
\fB\-b\fR, \fB\-\-base32\fR
use base32 encoding of KEY instead of hex
(default=off)
.TP
\fB\-c\fR, \fB\-\-counter\fR=\fI\,COUNTER\/\fR
HOTP counter value
.TP
\fB\-s\fR, \fB\-\-time\-step\-size\fR=\fI\,DURATION\/\fR TOTP time\-step duration
(default=`30s')
.TP
\fB\-S\fR, \fB\-\-start\-time\fR=\fI\,TIME\/\fR
when to start counting time steps for TOTP
(default=`1970\-01\-01 00:00:00 UTC')
.TP
\fB\-N\fR, \fB\-\-now\fR=\fI\,TIME\/\fR
use this time as current time for TOTP
(default=`now')
.TP
\fB\-d\fR, \fB\-\-digits\fR=\fI\,DIGITS\/\fR
number of digits in one\-time password
.TP
\fB\-w\fR, \fB\-\-window\fR=\fI\,WIDTH\/\fR
number of additional OTPs to generate or
validate against
.TP
\fB\-v\fR, \fB\-\-verbose\fR
explain what is being done  (default=off)
.SH EXAMPLES
To generate a time\-based (TOTP) one-time password from a key
protected by GnuPG:

  $ gpg \-\-decrypt \-\-quiet ~/.my-totp-secret.asc | oathtool \-\-totp \-

How to create the encrypted file is outside of scope of this manual,
see gpg(1), however the following two commands may help you to encrypt
the secret symmetrically (using a password) or asymmetrically (if you
have created a private key), respectively.  You will need to provide
the secret from the terminal after invoking the commands.

  $ gpg \-\-symmetric > ~/.my-totp-secret.asc
  $ gpg \-\-encrypt \-\-default-recipient-self > ~/.my-totp-secret.asc

For easier reading, the rest of this man page provides the KEY and OTP
on the command line.  Keep in mind that this usage is not recommended
since on most multi-user systems, command-line parameters of running
processes can be seen by other users.

To generate the first event-based (HOTP) one-time password for an all-zero key:

   $ echo 00 | oathtool \-
   328482
   $

Sometime you want to generate more than a single OTP.  To generate 10
additional event-based one-time passwords, with the secret key used in
the examples of RFC 4226, use the \-w (\-\-window) parameter:

   $ oathtool \-w 10 3132333435363738393031323334353637383930
   755224
   287082
   359152
   969429
   338314
   254676
   287922
   162583
   399871
   520489
   403154
   $

In the last output, the counter for the first OTP was 0, the second
OTP had a counter of 1, and so on up to 10.

In order to use keys encoded in Base32 instead of hex, you may provide
the \-b (\-\-base32) parameter:

   $ oathtool \-\-base32 \-w 3 GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ
   755224
   287082
   359152
   969429
   $

The tool ignore whitespace in base32 data and re-add padding if
necessary, thus you may supply keys formatted like the one below.

   $ oathtool \-\-base32 \-\-totp "gr6d 5br7 25s6 vnck v4vl hlao re"
   977872
   $

To generate a particular OTP, use the \-c (\-\-counter) parameter to
give the exact position directly:

   $ oathtool \-c 5 3132333435363738393031323334353637383930
   254676
   $

To validate a HOTP one-time password supply the OTP last on the
command line:

   $ oathtool \-w 10 3132333435363738393031323334353637383930 969429
   3
   $

The output indicates the counter that was used.  It works by starting
with counter 0 and increment until it founds a match (or not), within
the supplied window of 10 OTPs.

The tool supports time-variant one-time passwords, in so called TOTP
mode.  Usage is similar, but \-\-totp needs to be provided:

   $ oathtool \-\-totp 00
   943388
   $

Don't be alarmed if you do not get the same output, this is because
the output depends on the current time.  To generate a TOTP for a
particular fixed time use the \-N (\-\-now) parameter:

   $ oathtool \-\-totp \-\-now "2008\-04\-23 17:42:17 UTC" 00
   974945
   $

The format is a mostly free format human readable date string such as
"Sun, 29 Feb 2004 16:21:42 \-0800" or "2004\-02\-29 16:21:42" or even
"next Thursday".  It is the same used as the \-\-date parameter of the
date(1) tool.

The default MAC algorithm to use with TOTP is HMAC\-SHA1 and this is
what is usually used.  The tool supports two other MACs, namely the
HMAC\-SHA256 and HMAC\-SHA512 as well.  To use either of these,
qualify the \-\-totp parameter with a value.  Use "SHA256" for
HMAC\-SHA256 and "SHA512" for HMAC\-SHA512.  The following demonstrate
generating one of the RFC 6238 test vectors.

   $ oathtool \-\-totp=SHA256 \-\-digits=8 \-\-now "2009\-02\-13 23:31:30 UTC" 3132333435363738393031323334353637383930313233343536373839303132
   91819424
   $

You may generate several TOTPs by specifying the \-\-window parameter,
similar to how it works for HOTP.  The OTPs generated here will be for
the initial time (normally current time) and then each following time
step (e.g., 30 second window).

   $ oathtool \-\-totp 00 \-w5
   815120
   003818
   814756
   184042
   582326
   733842
   $

You can validate a TOTP one-time password by supplying the secret and
a window parameter (number of time steps before or after current
time):

   $ oathtool \-\-totp \-w 5 00 `oathtool \-\-totp 00`
   0
   $

Similar when generating TOTPs, you can use a \-N (\-\-now) parameter
to specify the time to use instead of the current time:

  $ oathtool \-\-totp \-\-now="2005\-03\-18 01:58:29 UTC" \-w 10000000 3132333435363738393031323334353637383930 89005924
  4115227
  $

The previous test uses values from the TOTP specification and will
stress test the tool because the expected window is around 4 million
time-steps.

There are two system parameters for TOTP: the time-step size and the
time start.

By default the time-step size is 30 seconds, which means you get a new
OTP every 30 seconds.  You may modify this with the \-s
(\-\-time\-step\-size) parameter:

   $ oathtool \-\-totp \-\-time-step-size=45s 00
   109841
   $

The values are valid ISO\-8601 durations, see:
http://en.wikipedia.org/wiki/ISO_8601#Durations

The time start is normally 1970\-01\-01 00:00:00 UTC but you may change
it using the \-S (\-\-start\-time):

   $ oathtool \-\-totp \-\-start-time "1980\-01\-01 00:00:00 UTC" 00
   273884
   $

To get more information about what the tool is using use the \-v
(\-\-verbose) parameter.  Finally, to generate the last TOTP (for
SHA\-1) in the test vector table of RFC 6238 you can invoke the tool
like this:

   $ oathtool \-\-totp \-v \-N "2033\-05\-18 03:33:20 UTC" \-d8 3132333435363738393031323334353637383930
   Hex secret: 3132333435363738393031323334353637383930
   Base32 secret: GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ
   Digits: 8
   Window size: 0
   TOTP mode: SHA1
   Step size (seconds): 30
   Start time: 1970\-01\-01 00:00:00 UTC (0)
   Time now: 2033\-05\-18 03:33:20 UTC (2000000000)
   Counter: 0x3F940AA (66666666)

   69279037
   $
.SH AUTHOR
Written by Simon Josefsson.
.SH "REPORTING BUGS"
Report bugs to: oath\-toolkit\-help@nongnu.org
oathtool home page: <https://www.nongnu.org/oath\-toolkit/>
.br
General help using GNU software: <https://www.gnu.org/gethelp/>
.SH COPYRIGHT
Copyright \(co 2009\-2024 Simon Josefsson.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


ft_otp/oath-toolkit-2.6.9/oathtool/oathtool.c
/*
 * oathtool.c - command line tool for OATH one-time passwords
 * Copyright (C) 2009-2023 Simon Josefsson
 *
 * This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */

#include <config.h>

#include "oath.h"

#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <inttypes.h>
#include <ctype.h>

/* Gnulib. */
#include "progname.h"
#include "error.h"
#include "version-etc.h"
#include "parse-duration.h"
#include "parse-datetime.h"

#include "oathtool_cmd.h"

const char version_etc_copyright[] =
  /* Do *not* mark this string for translation.  %s is a copyright
     symbol suitable for this locale, and %d is the copyright
     year.  */
  "Copyright %s 2009-%d Simon Josefsson.";

/* This feature is available in gcc versions 2.5 and later.  */
#if __GNUC__ < 2 || (__GNUC__ == 2 && __GNUC_MINOR__ < 5)
# define OATH_ATTR_NO_RETURN
#else
# define OATH_ATTR_NO_RETURN __attribute__ ((__noreturn__))
#endif

/* *INDENT-OFF* */
static void
usage (int status)
  OATH_ATTR_NO_RETURN;
/* *INDENT-ON* */

static void
usage (int status)
{
  if (status != EXIT_SUCCESS)
    fprintf (stderr, "Try `%s --help' for more information.\n", program_name);
  else
    {
      cmdline_parser_print_help ();
      emit_bug_reporting_address ();
    }
  exit (status);
}

static time_t
parse_time (const char *p, const time_t now)
{
  struct timespec nowspec = { 0, 0 };
  struct timespec thenspec;

  nowspec.tv_sec = now;

  if (!parse_datetime (&thenspec, p, &nowspec))
    return BAD_TIME;

  return thenspec.tv_sec;
}

static void
verbose_hotp (uint64_t moving_factor)
{
  printf ("Start counter: 0x%" PRIX64 " (%" PRIu64 ")\n\n",
	  moving_factor, moving_factor);
}

static void
verbose_totp (oath_totp_flags flags, time_t t0, time_t time_step_size,
	      time_t when)
{
  struct tm tmp;
  char outstr[200];

  printf ("TOTP mode: %s\n", flags == OATH_TOTP_HMAC_SHA256 ? "SHA256" :
	  flags == OATH_TOTP_HMAC_SHA512 ? "SHA512" :
	  flags == 0 ? "SHA1" : "invalid");

  if (gmtime_r (&t0, &tmp) == NULL)
    error (EXIT_FAILURE, 0, "gmtime_r");

  if (strftime (outstr, sizeof (outstr), "%Y-%m-%d %H:%M:%S UTC", &tmp) == 0)
    error (EXIT_FAILURE, 0, "strftime");

  printf ("Step size (seconds): %ld\n", time_step_size);
  printf ("Start time: %s (%ld)\n", outstr, t0);

  if (gmtime_r (&when, &tmp) == NULL)
    error (EXIT_FAILURE, 0, "gmtime_r");

  if (strftime (outstr, sizeof (outstr), "%Y-%m-%d %H:%M:%S UTC", &tmp) == 0)
    error (EXIT_FAILURE, 0, "strftime");

  printf ("Current time: %s (%ld)\n", outstr, when);
  printf ("Counter: 0x%lX (%ld)\n\n", (when - t0) / time_step_size,
	  (when - t0) / time_step_size);
}

#define generate_otp_p(n) ((n) == 1)
#define validate_otp_p(n) ((n) == 2)

/* Return newly allocated string with args_info->inputs[i] or, if that
   string starts with '-' or '@', a line read from stdin (-) or file
   (@...). */
const char *
maybe_read_input (const struct gengetopt_args_info *args_info, int i)
{
  const char *given;
  static FILE *f;
  static const char *last_file;
  char *lineptr = NULL;
  size_t n = 0;
  ssize_t l;

  if (i >= args_info->inputs_num)
    return NULL;

  given = args_info->inputs[i];

  if (*given != '-' && *given != '@')
    return strdup (given);

  if (!strcmp (given, "-"))
    f = stdin;
  else if (last_file == NULL || strcmp (given, last_file) != 0)
    {
      if (f)
	fclose (f);
      f = fopen (given + 1, "r");
      if (!f)
	error (EXIT_FAILURE, errno, "open file for KEY/OTP");
      last_file = given;
    }

  l = getline (&lineptr, &n, f);
  if (l < 0)
    error (EXIT_FAILURE, errno, "read file for KEY/OTP");
  if (l > 0 && lineptr[--l] == '\n')
    lineptr[l] = '\0';

  return lineptr;
}

#define EXIT_OTP_INVALID 2

int
main (int argc, char *argv[])
{
  struct gengetopt_args_info args_info;
  char *secret;
  size_t secretlen = 0;
  int rc;
  size_t window;
  uint64_t moving_factor;
  unsigned digits;
  char otp[10];
  time_t now, when, t0, time_step_size;
  int totpflags = 0;
  const char *inputs[2];

  set_program_name (argv[0]);

  if (cmdline_parser (argc, argv, &args_info) != 0)
    return EXIT_FAILURE;

  if (args_info.version_given)
    {
      char *p;
      int l = -1;

      if (strcmp (oath_check_version (NULL), OATH_VERSION) != 0)
	l = asprintf (&p, "OATH Toolkit %s/%s",
		      OATH_VERSION, oath_check_version (NULL));
      else if (strcmp (OATH_VERSION, PACKAGE_VERSION) != 0)
	l = asprintf (&p, "OATH Toolkit %s", OATH_VERSION);
      version_etc (stdout, "oathtool", l == -1 ? "OATH Toolkit" : p,
		   PACKAGE_VERSION, "Simon Josefsson", (char *) NULL);
      if (l != -1)
	free (p);
      return EXIT_SUCCESS;
    }

  if (args_info.help_given)
    usage (EXIT_SUCCESS);

  if (args_info.inputs_num == 0)
    {
      cmdline_parser_print_help ();
      emit_bug_reporting_address ();
      return EXIT_SUCCESS;
    }

  rc = oath_init ();
  if (rc != OATH_OK)
    error (EXIT_FAILURE, 0, "liboath initialization failed: %s",
	   oath_strerror (rc));

  inputs[0] = maybe_read_input (&args_info, 0);
  inputs[1] = maybe_read_input (&args_info, 1);

  if (args_info.base32_flag)
    {
      rc = oath_base32_decode (inputs[0],
			       strlen (inputs[0]), &secret, &secretlen);
      if (rc != OATH_OK)
	error (EXIT_FAILURE, 0, "base32 decoding failed: %s",
	       oath_strerror (rc));
    }
  else
    {
      secretlen = 1 + strlen (inputs[0]) / 2;
      secret = malloc (secretlen);
      if (!secret)
	error (EXIT_FAILURE, errno, "malloc");

      rc = oath_hex2bin (inputs[0], secret, &secretlen);
      if (rc != OATH_OK)
	error (EXIT_FAILURE, 0, "hex decoding of secret key failed");
    }

  if (args_info.counter_orig)
    moving_factor = args_info.counter_arg;
  else
    moving_factor = 0;

  if (args_info.digits_orig)
    digits = args_info.digits_arg;
  else
    digits = 6;

  if (args_info.window_orig)
    window = args_info.window_arg;
  else
    window = 0;

  if (digits != 6 && digits != 7 && digits != 8)
    error (EXIT_FAILURE, 0, "only digits 6, 7 and 8 are supported");

  if (validate_otp_p (args_info.inputs_num) && !args_info.digits_orig)
    digits = strlen (inputs[1]);
  else if (validate_otp_p (args_info.inputs_num) && args_info.digits_orig &&
	   args_info.digits_arg != strlen (inputs[1]))
    error (EXIT_FAILURE, 0,
	   "given one-time password has bad length %d != %ld",
	   args_info.digits_arg, strlen (inputs[1]));

  if (args_info.inputs_num > 2)
    error (EXIT_FAILURE, 0, "too many parameters");

  if (args_info.verbose_flag)
    {
      char *tmp;

      tmp = malloc (2 * secretlen + 1);
      if (!tmp)
	error (EXIT_FAILURE, errno, "malloc");

      oath_bin2hex (secret, secretlen, tmp);

      printf ("Hex secret: %s\n", tmp);
      free (tmp);

      rc = oath_base32_encode (secret, secretlen, &tmp, NULL);
      if (rc != OATH_OK)
	error (EXIT_FAILURE, 0, "base32 encoding failed: %s",
	       oath_strerror (rc));

      printf ("Base32 secret: %s\n", tmp);
      free (tmp);

      if (args_info.inputs_num == 2)
	printf ("OTP: %s\n", inputs[1]);
      printf ("Digits: %d\n", digits);
      printf ("Window size: %ld\n", window);
    }

  if (args_info.totp_given)
    {
      char *p;

      now = time (NULL);
      when = parse_time (args_info.now_arg, now);
      t0 = parse_time (args_info.start_time_arg, now);
      time_step_size = parse_duration (args_info.time_step_size_arg);

      if (when == BAD_TIME)
	error (EXIT_FAILURE, 0, "cannot parse time `%s'", args_info.now_arg);

      if (t0 == BAD_TIME)
	error (EXIT_FAILURE, 0, "cannot parse time `%s'",
	       args_info.start_time_arg);

      if (time_step_size == BAD_TIME)
	error (EXIT_FAILURE, 0, "cannot parse time `%s'",
	       args_info.time_step_size_arg);

      for (p = args_info.totp_arg; *p; p++)
	*p = toupper (*p);
      if (strcmp (args_info.totp_arg, "SHA256") == 0)
	totpflags = OATH_TOTP_HMAC_SHA256;
      else if (strcmp (args_info.totp_arg, "SHA512") == 0)
	totpflags = OATH_TOTP_HMAC_SHA512;

      if (args_info.verbose_flag)
	verbose_totp (totpflags, t0, time_step_size, when);
    }
  else
    {
      if (args_info.verbose_flag)
	verbose_hotp (moving_factor);
    }

  if (generate_otp_p (args_info.inputs_num) && !args_info.totp_given)
    {
      size_t iter = 0;

      do
	{
	  rc = oath_hotp_generate (secret,
				   secretlen,
				   moving_factor + iter,
				   digits,
				   false, OATH_HOTP_DYNAMIC_TRUNCATION, otp);
	  if (rc != OATH_OK)
	    error (EXIT_FAILURE, 0,
		   "generating one-time password failed (%d)", rc);

	  printf ("%s\n", otp);
	}
      while (window - iter++ > 0);
    }
  else if (generate_otp_p (args_info.inputs_num) && args_info.totp_given)
    {
      size_t iter = 0;

      do
	{
	  rc = oath_totp_generate2 (secret,
				    secretlen,
				    when + iter * time_step_size,
				    time_step_size, t0, digits, totpflags,
				    otp);
	  if (rc != OATH_OK)
	    error (EXIT_FAILURE, 0,
		   "generating one-time password failed (%d)", rc);

	  printf ("%s\n", otp);
	}
      while (window - iter++ > 0);
    }
  else if (validate_otp_p (args_info.inputs_num) && !args_info.totp_given)
    {
      rc = oath_hotp_validate (secret,
			       secretlen, moving_factor, window, inputs[1]);
      if (rc == OATH_INVALID_OTP)
	error (EXIT_OTP_INVALID, 0,
	       "password \"%s\" not found in range %ld .. %ld",
	       inputs[1],
	       (long) moving_factor, (long) moving_factor + window);
      else if (rc < 0)
	error (EXIT_FAILURE, 0,
	       "validating one-time password failed (%d)", rc);
      printf ("%d\n", rc);
    }
  else if (validate_otp_p (args_info.inputs_num) && args_info.totp_given)
    {
      rc = oath_totp_validate4 (secret,
				secretlen,
				when,
				time_step_size,
				t0, window, NULL, NULL, totpflags, inputs[1]);
      if (rc == OATH_INVALID_OTP)
	error (EXIT_OTP_INVALID, 0,
	       "password \"%s\" not found in range %ld .. %ld",
	       inputs[1],
	       (long) ((when - t0) / time_step_size - window / 2),
	       (long) ((when - t0) / time_step_size + window / 2));
      else if (rc < 0)
	error (EXIT_FAILURE, 0,
	       "validating one-time password failed (%d)", rc);
      printf ("%d\n", rc);
    }

  free (secret);
  oath_done ();

  return EXIT_SUCCESS;
}
